R/htmlwidgets.R

```diff
@@ -228,18 +228,54 @@ toHTML <- function(x, standalone = FALSE, knitrOptions = NULL) {
 
 }
 
+widget_html_wrapper <- function(name, package, id, style, class, inline = FALSE, ...) {
+
+  fn_res <- widget_html(
+    name = structure(name, class = c("character", name)),
+    package = package,
+    id = id,
+    style = style,
+    class = class,
+    inline = inline,
+    ...)
+
+  if(!inherits(fn_res,c("shiny.tag","shiny.tag.list"))){
+    warning(
+      paste0("widget_html.",name),
+      " returned an object of class `",
+      deparse(class(fn_res)),
+      "` instead of a `shiny.tag`."
+    )
+  }
+
+  fn_res
+
+}
 
-widget_html <- function(name, package, id, style, class, inline = FALSE, ...){
+widget_html <- function (name, package, id, style, class, inline = FALSE, ...) {
+  UseMethod("widget_html", name)
+}
+
+
+widget_html.default <- function (name, package, id, style, class, inline = FALSE, ...) {
 
-  # attempt to lookup custom html function for widget
   fn <- tryCatch(get(paste0(name, "_html"),
                      asNamespace(package),
                      inherits = FALSE),
                  error = function(e) NULL)
 
-  # call the custom function if we have one, otherwise create a div
   if (is.function(fn)) {
+    .Deprecated(
+      new = paste0("widget_html.",name),
+      package = package,
+      msg = paste("Using `widgetname_html` is about to be deprecated.",
+                  "widget_html going forward will now follow standard s3 method",
+                  "deployment for creating custom HTML 'housings'.",
+                  "See docs for details at http://www.htmlwidgets.org/develop_advanced.html"),
+      old = paste0(name, "_html")
+    )
     fn(id = id, style = style, class = class, ...)
+
```

Comment:
Can you please move the "x returned an object of class y instead of shiny.tag" check/warning to here? I'd rather not check the results if they've implemented the S3 method, as there are conceivably other things you could return as housing besides tags (`HTML()`, objects coercible using `htmltools::as.tags`).

Reply:
sure thing